openmc/weight_windows.py

```diff
@@ -410,3 +406,180 @@ def from_hdf5(cls, group, meshes):
+
+    # create openmc mesh object
+    grids = []
+    mesh_definition = [(x0, x_vals, nfx), (y0, y_vals, nfy), (z0, z_vals, nfz)]
+    for grid0, grid_vals, n_pnts in mesh_definition:
+        # file spec checks for the mesh definition
+        if not all(grid_vals[2::3] == 1.0):
+            raise ValueError('One or more mesh ratio value, qx, '
+                             'is not equal to one')
+
+        if np.sum(grid_vals[::3]) != n_pnts:
+            raise ValueError('Sum of the fine bin entries, s, does '
+                             'not match the number of fine bins')
+
+        # extend the grid based on the next coarse bin endpoint, px
+        # and the number of fine bins in the coarse bin, sx
+        intervals = grid_vals.reshape(-1, 3)
+        coords = [grid0]
+        for sx, px, qx in intervals:
+            coords += list(np.linspace(coords[-1], px, int(sx + 1)))[1:]
+
+        grids.append(np.array(coords))
+
+    mesh = RectilinearMesh()
+    mesh.x_grid, mesh.y_grid, mesh.z_grid = grids
+
+    # extract weight window values from array
+    particle_types = {0: 'neutron', 1: 'photon'}
+    particle_data = []
+    for p in range(ni):
+        # no information to read for this particle if
+        # either the energy bins or time bins are empty
+        if ne[p] == 0 or nt[p] == 0:
+            continue
+
+        if iv > 1:
+            # time bins are parsed but unused for now
+            end_idx = start_idx + nt[p]
+            time_bounds = ww_data[start_idx:end_idx]
+            np.insert(time_bounds, (0,), (0.0,))
+            start_idx = end_idx
+
+        # read energy boundaries
+        end_idx = start_idx + ne[p]
+        energy_bounds = np.insert(ww_data[start_idx:end_idx], (0,), (0.0,))
+        # convert from MeV to eV
+        energy_bounds *= 1e6
+        start_idx = end_idx
+
+        # read weight window values
+        end_idx = start_idx + (nfx * nfy * nfz) * nt[p] * ne[p]
+
+        # read values and reshape according to ordering
+        # slowest to fastest: z, y, x, e
+        ww_values = ww_data[start_idx:end_idx].reshape(nfz, nfy, nfx, ne[p])
+        # swap z and x axes for correct shape and (ijk, e) mesh indexing
+        ww_values = np.swapaxes(ww_values, 2, 0)
+        start_idx = end_idx
+
+        # gather particle data in container
+        particle_dict = {'particle': particle_types[p],
+                         'energy_bounds': energy_bounds,
+                         'ww_values': ww_values}
+        particle_data.append(particle_dict)
+
+    # create openmc weight window objects
+    wws = []
+    for data in particle_data:
+        ww = WeightWindows(id=None,
+                           mesh=mesh,
+                           lower_ww_bounds=data['ww_values'],
+                           upper_bound_ratio=5.0,
+                           energy_bounds=data['energy_bounds'],
+                           particle_type=data['particle'])
+        wws.append(ww)
```

Comment:
Could this just be combined with the previous loop?

Reply:
For sure. Good call.